File: docs/core/examples/ruby/assets.rb
require 'chain'

chain = Chain::Client.new
signer = Chain::HSMSigner.new

asset_key = chain.mock_hsm.keys.create
signer.add_key(asset_key, chain.mock_hsm.signer_conn)

account_key = chain.mock_hsm.keys.create
signer.add_key(account_key, chain.mock_hsm.signer_conn)

chain.accounts.create(
  alias: 'acme_treasury',
  root_xpubs: [account_key.xpub],
  quorum: 1,
)

# snippet create-asset-acme-common
chain.assets.create(
  alias: 'acme_common',
  root_xpubs: [asset_key.xpub],
  quorum: 1,
  tags: {
    internal_rating: '1',
  },
  definition: {
    issuer: 'Acme Inc.',
    type: 'security',
    subtype: 'private',
    class: 'common',
  },
)
# endsnippet

# snippet create-asset-acme-preferred
chain.assets.create(
  alias: 'acme_preferred',
  root_xpubs: [asset_key.xpub],
  quorum: 1,
  tags: {
    internal_rating: '2',
  },
  definition: {
    issuer: 'Acme Inc.',
    type: 'security',
    subtype: 'private',
    class: 'preferred',
  },
)
# endsnippet

# snippet list-local-assets
chain.assets.query(
  filter: 'is_local=$1',
  filter_params: ['yes'],
).each do |a|
  puts "Local asset: #{a.alias}"
end
# endsnippet

# snippet list-private-preferred-securities
chain.assets.query(
  filter: 'definition.type=$1 AND definition.subtype=$2 AND definition.class=$3',
  filter_params: ['security', 'private', 'preferred'],
).each do |a|
  puts "Private preferred security: #{a.alias}"
end
# endsnippet

# snippet build-issue
issuance_tx = chain.transactions.build do |b|
  b.issue asset_alias: 'acme_common', amount: 1000
  b.control_with_account account_alias: 'acme_treasury', asset_alias: 'acme_common', amount: 1000
end
# endsnippet

# snippet sign-issue
signed_issuance_tx = signer.sign(issuance_tx)
# endsnippet

# snippet submit-issue
chain.transactions.submit(signed_issuance_tx)
# endsnippet

external_program = chain.accounts.create_control_program(
  alias: 'acme_treasury'
).control_program

# snippet external-issue
external_issuance = chain.transactions.build do |b|
  b.issue asset_alias: 'acme_preferred', amount: 2000
  b.control_with_program control_program: external_program, asset_alias: 'acme_preferred', amount: 2000
end

chain.transactions.submit(signer.sign(external_issuance))
# endsnippet

# snippet build-retire
retirement_tx = chain.transactions.build do |b|
  b.spend_from_account account_alias: 'acme_treasury', asset_alias: 'acme_common', amount: 50
  b.retire asset_alias: 'acme_common', amount: 50
end
# endsnippet

# snippet sign-retire
signed_retirement_tx = signer.sign(retirement_tx)
# endsnippet

# snippet submit-retire
chain.transactions.submit(signed_retirement_tx)
# endsnippet

# snippet list-issuances
chain.transactions.query(
  filter: 'inputs(type=$1 AND asset_alias=$2)',
  filter_params: ['issue', 'acme_common'],
).each do |t|
  puts "Acme Common issued in tx #{t.id}"
end
# endsnippet

# snippet list-transfers
chain.transactions.query(
  filter: 'inputs(type=$1 AND asset_alias=$2)',
  filter_params: ['spend', 'acme_common'],
).each do |t|
  puts "Acme Common transferred in tx #{t.id}"
end
# endsnippet

# snippet list-retirements
chain.transactions.query(
  filter: 'outputs(type=$1 AND asset_alias=$2)',
  filter_params: ['retire', 'acme_common'],
).each do |t|
  puts "Acme Common retired in tx #{t.id}"
end
# endsnippet

# snippet list-acme-common-balance
chain.balances.query(
  filter: 'asset_alias=$1',
  filter_params: ['acme_common'],
).each do |b|
  puts "Total circulation of Acme Common: #{b.amount}"
end
# endsnippet

# snippet list-acme-balance
chain.balances.query(
  filter: 'asset_definition.issuer=$1',
  filter_params: ['Acme Inc.'],
).each do |b|
  puts "Total circulation of Acme stock #{b.sum_by['asset_alias']}: #{b.amount}"
end
# endsnippet

# snippet list-acme-common-unspents
chain.unspent_outputs.query(
  filter: 'asset_alias=$1',
  filter_params: ['acme_common'],
).each do |u|
  puts "Acme Common held in output #{u.transaction_id}:#{u.position}"
end
# endsnippet

File: docs/core/examples/node/assets.js
const chain = require('chain-sdk')

const client = new chain.Client()
const signer = new chain.HsmSigner()
let assetKey, accountKey

Promise.all([
  client.mockHsm.keys.create(),
  client.mockHsm.keys.create(),
]).then(keys => {
  assetKey = keys[0].xpub
  accountKey = keys[1].xpub

  signer.addKey(assetKey, client.mockHsm.signerConnection)
  signer.addKey(accountKey, client.mockHsm.signerConnection)
}).then(() => Promise.all([
  client.accounts.create({
    alias: 'acme_treasury',
    rootXpubs: [accountKey],
    quorum: 1,
  }),

  // snippet create-asset-acme-common
  client.assets.create({
    alias: 'acme_common',
    rootXpubs: [assetKey],
    quorum: 1,
    tags: {
      internalRating: '1',
    },
    definition: {
      issuer: 'Acme Inc.',
      type: 'security',
      subtype: 'private',
      class: 'common',
    },
  })
  // endsnippet
  ,
  // snippet create-asset-acme-preferred
  client.assets.create({
    alias: 'acme_preferred',
    rootXpubs: [assetKey],
    quorum: 1,
    tags: {
      internalRating: '2',
    },
    definition: {
      issuer: 'Acme Inc.',
      type: 'security',
      subtype: 'private',
      class: 'preferred',
    },
  })
  // endsnippet
])).then(() =>
  // snippet list-local-assets
  client.assets.queryAll({
    filter: 'is_local=$1',
    filterParams: ['yes'],
  }, (asset, next) => {
    console.log('Local asset: ' + asset.alias)
    next()
  })
  // endsnippet
).then(() =>
  // snippet list-private-preferred-securities
  client.assets.queryAll({
    filter: 'definition.type=$1 AND definition.subtype=$2 AND definition.class=$3',
    filterParams: ['security', 'private', 'preferred'],
  }, (asset, next) => {
    console.log('Private preferred security: ' + asset.alias)
    next()
  })
  // endsnippet
).then(() => {
  // snippet build-issue
  const issuePromise = client.transactions.build(builder => {
    builder.issue({
      assetAlias: 'acme_common',
      amount: 1000
    })
    builder.controlWithAccount({
      accountAlias: 'acme_treasury',
      assetAlias: 'acme_common',
      amount: 1000
    })
  })
  // endsnippet

  return issuePromise.then(issueTx => {
    // snippet sign-issue
    const signingPromise = signer.sign(issueTx)
    // endsnippet

    return signingPromise
  }).then(signedIssueTx =>
    // snippet submit-issue
    client.transactions.submit(signedIssueTx)
    // endsnippet
  )
}).then(() => {
  const externalReceiverPromise = client.accounts.createReceiver({
    accountAlias: 'acme_treasury',
  })

  return externalReceiverPromise.then(externalReceiver =>
    // snippet external-issue
    client.transactions.build(builder => {
      builder.issue({
        assetAlias: 'acme_preferred',
        amount: 2000
      })
      builder.controlWithReceiver({
        receiver: externalReceiver,
        assetAlias: 'acme_preferred',
        amount: 2000
      })
    }).then(template => {
      return signer.sign(template)
    }).then(signed => {
      return client.transactions.submit(signed)
    })
    // endsnippet
  )
}).then(() => {
  // snippet build-retire
  const retirePromise = client.transactions.build(builder => {
    builder.spendFromAccount({
      accountAlias: 'acme_treasury',
      assetAlias: 'acme_common',
      amount: 50
    })
    builder.retire({
      assetAlias: 'acme_common',
      amount: 50
    })
  })
  // endsnippet

  return retirePromise.then(retireTx => {
    // snippet sign-retire
    const signingPromise = signer.sign(retireTx)
    // endsnippet

    return signingPromise
  }).then(signedRetireTx =>
    // snippet submit-retire
    client.transactions.submit(signedRetireTx)
    // endsnippet
  )
}).then(() =>
  // snippet list-issuances
  client.transactions.queryAll({
    filter: 'inputs(type=$1 AND asset_alias=$2)',
    filterParams: ['issue', 'acme_common'],
  }, (tx, next) => {
    console.log('Acme Common issued in tx ' + tx.id)
    next()
  })
  // endsnippet
).then(() =>
  // snippet list-transfers
  client.transactions.queryAll({
    filter: 'inputs(type=$1 AND asset_alias=$2)',
    filterParams: ['spend', 'acme_common'],
  }, (tx, next) => {
    console.log('Acme Common transferred in tx ' + tx.id)
    next()
  })
  // endsnippet
).then(() =>
  // snippet list-retirements
  client.transactions.queryAll({
    filter: 'outputs(type=$1 AND asset_alias=$2)',
    filterParams: ['retire', 'acme_common'],
  }, (tx, next) => {
    console.log('Acme Common retired in tx ' + tx.id)
    next()
  })
  // endsnippet
).then(() =>
  // snippet list-acme-common-balance
  client.balances.queryAll({
    filter: 'asset_alias=$1',
    filterParams: ['acme_common'],
  }, (balance, next) => {
    console.log('Total circulation of Acme Common: ' + balance.amount)
    next()
  })
  // endsnippet
).then(() =>
  // snippet list-acme-balance
  client.balances.queryAll({
    filter: 'asset_definition.issuer=$1',
    filterParams: ['Acme Inc.'],
  }, (balance, next) => {
    console.log('Total circulation of Acme stock ' + balance.sumBy.assetAlias + ': ' + balance.amount)
    next()
  })
  // endsnippet
).then(() =>
  // snippet list-acme-common-unspents
  client.unspentOutputs.queryAll({
    filter: 'asset_alias=$1',
    filterParams: ['acme_common'],
  }, (unspent, next) => {
    console.log('Acme Common held in output ' + unspent.id)
    next()
  })
  // endsnippet
).catch(err =>
  process.nextTick(() => { throw err })
)
